File: shared/src/main/java/ru/privatenull/pnauth/limbo/LimboServerRegistry.java
package ru.privatenull.pnauth.limbo;

import java.util.LinkedHashMap;
import java.util.Map;

public final class LimboServerRegistry {
    private final Map<String, LimboServerProvider> providers = new LinkedHashMap<>();

    public void register(LimboServerProvider provider) {
        if (provider == null || provider.id() == null || provider.id().isBlank()) {
            throw new IllegalArgumentException("Invalid limbo provider");
        }
        if (providers.putIfAbsent(provider.id().toLowerCase(), provider) != null) {
            throw new IllegalArgumentException("Limbo provider is already registered: " + provider.id());
        }
    }

    public LimboServer create(String providerId, LimboServerContext context) {
        LimboServerProvider provider = providers.get(providerId.toLowerCase());
        if (provider == null) {
            throw new IllegalArgumentException("Unknown limbo provider: " + providerId);
        }
        return provider.create(context);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/limbo/LimboServerRegistry.kt
package ru.privatenull.pnauth.limbo

import java.util.LinkedHashMap
import java.util.Locale

class LimboServerRegistry {
    private val providers = LinkedHashMap<String, LimboServerProvider>()

    fun register(provider: LimboServerProvider) {
        require(provider.id().isNotBlank()) { "Invalid limbo provider" }
        val key = provider.id().lowercase(Locale.ROOT)
        require(!providers.containsKey(key)) { "Limbo provider is already registered: ${provider.id()}" }
        providers[key] = provider
    }

    fun create(providerId: String, context: LimboServerContext): LimboServer {
        val provider = providers[providerId.lowercase(Locale.ROOT)]
            ?: throw IllegalArgumentException("Unknown limbo provider: $providerId")
        return provider.create(context)
    }
}
